### PR TITLE
Implement to_constr with nf_evars_and_universes_opt_subst

### PR DESCRIPTION
**Pull request by SkySkimmer**
Required fixing nf_evars_universes_opt_subst to recurse on universes (ie if the universe substitution is u := v := w using it on u will return w) and to normalise arguments of undefined evars.

This makes evar normalisation cbv in the sense that evar arguments are normalised before substituting them in the evar's body.
Also to_constr used the self-updating univ normalisation functions but didn't share the ref, ie
```ocaml
let univ1 = nf (ref subst) univ1 and univ2 = nf (ref subst) univ2
```
instead of what we now to
```ocaml
let nf = nf (ref subst) in
let univ1 = nf univ1 and univ2 = nf univ2
```

Bench (before rebase): https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/404/console
```
┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1408.05 1440.90 -2.28 % │  3912069802801  4003762861640 -2.29 % │  6665731953227  6646604539069 +0.29 % │ 1407780 1403920 +0.27 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-hott │  329.34  336.62 -2.16 % │   913233983757   933379177639 -2.16 % │  1426110059064  1426839554879 -0.05 % │  588964  588828 +0.02 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │  261.42  264.24 -1.07 % │   725156743938   732823670709 -1.05 % │   903748177449   899331178619 +0.49 % │ 1245532 1245380 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   79.60   80.41 -1.01 % │   219418897712   221079278372 -0.75 % │   285573792696   285690661579 -0.04 % │  528284  528652 -0.07 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  868.44  876.86 -0.96 % │  2405984392941  2429431867205 -0.97 % │  3482722749416  3476006758268 +0.19 % │ 1333280 1329108 +0.31 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.48   37.79 -0.82 % │   100995981889   101282318591 -0.28 % │   126047699020   126195986126 -0.12 % │  482408  482436 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-unimath │ 1276.08 1283.38 -0.57 % │  3530063936629  3548092059022 -0.51 % │  5999502936648  6004230064763 -0.08 % │ 1076892 1091456 -1.33 % │  39   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  274.78  276.11 -0.48 % │   762219544388   765373643004 -0.41 % │  1087135793433  1085630982712 +0.14 % │ 1120884 1131588 -0.95 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   65.00   65.26 -0.40 % │   178842486852   178414899422 +0.24 % │   239174205704   238307838259 +0.36 % │  589704  590880 -0.20 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-math-classes │  228.16  228.82 -0.29 % │   627744868219   627229854102 +0.08 % │   786027262674   784173178808 +0.24 % │  531944  532668 -0.14 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  201.44  201.89 -0.22 % │   557580914640   558244912316 -0.12 % │   769586593420   768272292958 +0.17 % │  865676  867160 -0.17 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   45.11   45.20 -0.20 % │   123296094719   123316644372 -0.02 % │   145032757636   144769805649 +0.18 % │  537660  536668 +0.18 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 3833.82 3839.96 -0.16 % │ 10658797158805 10675768553195 -0.16 % │ 14040496095441 14009721670617 +0.22 % │ 2226568 2221092 +0.25 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  700.30  700.82 -0.07 % │  1939534684572  1940974586444 -0.07 % │  2978135195707  2969479953820 +0.29 % │ 3353896 3375304 -0.63 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.51  111.56 -0.04 % │   310269950095   310361103773 -0.03 % │   379515541221   379805117973 -0.08 % │  501536  500772 +0.15 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3104.69 3103.00 +0.05 % │  8614152702313  8612484321462 +0.02 % │ 13820050492886 13789210766086 +0.22 % │ 1211548 1211572 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  172.49  172.34 +0.09 % │   477834343848   477789219408 +0.01 % │   698446289256   697778453135 +0.10 % │  837524  838028 -0.06 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  185.85  185.68 +0.09 % │   513287794509   513696513328 -0.08 % │   692732921118   692000074199 +0.11 % │  645200  643556 +0.26 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  723.98  722.63 +0.19 % │  2000695720428  1997628907142 +0.15 % │  2385209213298  2379504231054 +0.24 % │ 1446144 1446156 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  461.44  460.31 +0.25 % │  1280740406909  1277737394522 +0.24 % │  2064119015143  2060568818285 +0.17 % │  814348  814316 +0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   81.77   81.37 +0.49 % │   225460956306   225326142427 +0.06 % │   267533553131   266965646535 +0.21 % │  718368  717956 +0.06 % │  16   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1570.29 1561.85 +0.54 % │  4345543981915  4319529473941 +0.60 % │  6450177839995  6410983923008 +0.61 % │  902124  864792 +4.32 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘

```

### CODE REVIEW

File: engine/evd.ml [hunk 1]
```diff
@@ -1290,30 +1290,14 @@ module MiniEConstr = struct
...
-        let u' = normalize_universe_instance sigma u in
-        if u' == u then c else mkIndU (i, u')
-      | Construct (co, u) when not (Univ.Instance.is_empty u) ->
-        let u' = normalize_universe_instance sigma u in
-        if u' == u then c else mkConstructU (co, u')
-      | _ -> Constr.map (fun c -> to_constr c) c
-    in to_constr c
+    let evar_value =
+      if not abort_on_undefined_evars then safe_evar_value sigma
+      else fun ev ->
```
Comment (ppedrot) [inline engine/evd.ml]: In absence of flambda, it is usually advised against creating closures with an arity mismatch in performance intensive code. I think it'd be better to eta-expand the first branch of the if clause as `fun ev -> safe_evar_value sigma ev`.

**Review by ppedrot — COMMENTED** [review]
Did you also try a call-by-need reduction? Most of the evar arguments are never used in practice, they tend to be types for the actually used arguments. It might be interesting to see if that's any better.

File: engine/evd.ml (hunk 1 above)
Comment (SkySkimmer) [reply engine/evd.ml]: Changed.

**Comment by SkySkimmer** [comment]
>Did you also try a call-by-need reduction? 

https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/416/console (vs this PR)

**Review by ppedrot — APPROVED** [review]
I'm a bit surprised, but the call-by-need variant is actually slower than the call-by-value one. I'm OK to merge this PR as-is then.

**Comment by SkySkimmer** [comment]
Yup
```

┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.24   37.50 -0.69 % │   101266223885   101271594742 -0.01 % │   126138101215   126258361889 -0.10 % │  482336  482392 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   65.15   65.38 -0.35 % │   178839255697   179129546643 -0.16 % │   238950748530   239495366170 -0.23 % │  588620  587724 +0.15 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.50  111.78 -0.25 % │   310359926016   310581351423 -0.07 % │   380140793977   379703308496 +0.12 % │  501484  499996 +0.30 % │   6   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   45.33   45.40 -0.15 % │   123619957829   123684693354 -0.05 % │   145279122193   145253662037 +0.02 % │  537388  537100 +0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-color │  723.10  723.98 -0.12 % │  2003126089499  2004692282009 -0.08 % │  2385584015307  2384983084752 +0.03 % │ 1446524 1446548 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1573.22 1574.03 -0.05 % │  4368182992158  4369277697495 -0.03 % │  6453097402236  6449445545067 +0.06 % │  873980  874404 -0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real-closed │  171.96  172.00 -0.02 % │   476652224055   476915844849 -0.06 % │   698823540715   698528112150 +0.04 % │  838776  837448 +0.16 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  459.76  459.72 +0.01 % │  1276009361341  1276436984663 -0.03 % │  2064457531031  2064472722311 -0.00 % │  813316  814388 -0.13 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 3832.46 3831.12 +0.03 % │ 10656079224666 10655973572239 +0.00 % │ 14002309299214 14000553189431 +0.01 % │ 2234212 2234724 -0.02 % │   2   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  186.24  186.08 +0.09 % │   515518888511   513897874625 +0.32 % │   693832062748   693234231273 +0.09 % │  652652  644904 +1.20 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   80.06   79.90 +0.20 % │   220654756353   219712025923 +0.43 % │   285701151878   285988909546 -0.10 % │  528548  528616 -0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  703.65  702.17 +0.21 % │  1951317595424  1944755196060 +0.34 % │  2980413134201  2979995719706 +0.01 % │ 3375476 3376236 -0.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  276.56  275.70 +0.31 % │   766344689860   764499728308 +0.24 % │  1088434287598  1087312712092 +0.10 % │ 1132872 1121436 +1.02 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3097.02 3083.40 +0.44 % │  8613029875331  8572493930041 +0.47 % │ 13836134491087 13814022084740 +0.16 % │ 1239784 1239816 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  203.33  202.42 +0.45 % │   561958395324   559392394758 +0.46 % │   773529501350   770727817386 +0.36 % │  841616  837464 +0.50 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-unimath │ 1284.45 1278.29 +0.48 % │  3550440344263  3536801519039 +0.39 % │  6005076578706  6001027090093 +0.07 % │ 1079020 1078248 +0.07 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   82.14   81.64 +0.61 % │   226239909879   226190158756 +0.02 % │   267841005389   267652463716 +0.07 % │  720108  716868 +0.45 % │  21   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd-order │ 1434.70 1425.67 +0.63 % │  3986762402192  3961695177403 +0.63 % │  6673827869924  6665083810057 +0.13 % │ 1428084 1381592 +3.37 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-hott │  334.26  331.72 +0.77 % │   927615762003   921816288890 +0.63 % │  1426208476055  1426312732174 -0.01 % │  607864  589048 +3.19 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-math-classes │  231.26  229.36 +0.83 % │   634473294121   630089271368 +0.70 % │   790216629729   786228587486 +0.51 % │  531992  533328 -0.25 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │  264.57  262.09 +0.95 % │   735400108870   728019848516 +1.01 % │   903388040303   903507991881 -0.01 % │ 1245464 1245244 +0.02 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  878.54  868.90 +1.11 % │  2433591568552  2409792264709 +0.99 % │  3483311767031  3482453409814 +0.02 % │ 1334800 1333148 +0.12 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘

```

**Pull request merged**